editor/src/marionette/preview.ts

```diff
@@ -0,0 +1,357 @@
+import { BlendStateBuffer, LegacyBlendStateBuffer } from '../../../cocos/3d/skeletal-animation/skeletal-animation-blending';
+import { Node } from '../../../cocos/core';
+import { Motion, MotionEval } from '../../../cocos/core/animation/marionette/motion';
+import { createEval } from '../../../cocos/core/animation/marionette/create-eval';
+import { BindContext } from '../../../cocos/core/animation/marionette/parametric';
+import { VarInstance, Value, VariableType } from '../../../cocos/core/animation/marionette/variable';
+import { assertIsNonNullable } from '../../../cocos/core/data/utils/asserts';
+import {
+    AnimationBlendEval,
+} from '../../../cocos/core/animation/marionette/animation-blend';
+import type { RuntimeID } from '../../../cocos/core/animation/marionette/graph-debug';
+
+class AnimationGraphPartialPreviewer {
+    constructor(root: Node) {
+        this._root = root;
+    }
+
+    public destroy() {
+    }
+
+    public evaluate() {
+        this._blendBuffer.apply();
+    }
+
+    public addVariable(id: string, type: VariableType, value: Value) {
+        this._varInstances[id] = new VarInstance(type, value);
+    }
+
+    public removeVariable(id: string) {
+        delete this._varInstances[id];
+    }
+
+    public updateVariable(id: string, value: Value) {
+        this._varInstances[id].value = value;
+    }
+
+    protected createMotionEval(motion: Motion) {
+        const {
+            _root: root,
+            _blendBuffer: blendBuffer,
+        } = this;
+
+        const bindContext: BindContext = {
+            getVar: this._getVar.bind(this),
+        };
+
+        const motionEval = motion[createEval]({
+            node: root,
+            blendBuffer,
+            getVar: (...args) => {
+                return bindContext.getVar(...args);
+            },
+        });
+
+        return motionEval;
+    }
+
+    private _root: Node;
+
+    private _blendBuffer: BlendStateBuffer = new LegacyBlendStateBuffer();
+
+    private _varInstances: Record<string, VarInstance> = {};
+
+    private _getVar(id: string): VarInstance | undefined {
+        return this._varInstances[id];
+    }
+}
+
+export interface MotionPreviewerTimelineStats {
+    timeLineLength: number;
+}
+
+export class MotionPreviewer extends AnimationGraphPartialPreviewer {
+    get timelineStats(): Readonly<MotionPreviewerTimelineStats> {
+        if (this._timelineStatsDirty) {
+            this._updateTimelineStats();
+            this._timelineStatsDirty = false;
+        }
+        return this._timelineStats;
+    }
+
+    /**
+     * Gets an iterable to the weights of each motion(that has runtime ID).
+     */
+    public queryWeights(): Iterable<[RuntimeID, number]> {
+        const getWeightsRecursive = function* (motionEval: MotionEval, weight: number): Iterable<[RuntimeID, number]> {
```

Comment:
What does function* mean?

Reply:
It's a generator whose usage for example is:

```ts
const [[id, weight]] = queryWeights(); // Get the first motion's id and weight
for (const [id, weight] of queryWeights()) { // Iterate motion weights
  if (id === 'a') {
     break;
  }
}
```

Generators save us from array construction/concatenation e.g.